libs-haskell/da-hs-base/src/DA/Service/Logger/Impl/GCP.hs

```diff
@@ -399,16 +403,17 @@ today = Time.utctDay <$> getCurrentTime
 maxDataPerDay :: SentBytes
 maxDataPerDay = SentBytes (8 * 2 ^ (20 :: Int))
 
--- | The DAML home folder, getting this ensures the folder exists
-getDamlDir :: Maybe FilePath -> IO FilePath
-getDamlDir damlPathM = do
-    dh <- lookupEnv damlPathEnvVar
-    dir <- case damlPathM <|> dh of
+-- | The DAML cache folder, getting this ensures the folder exists.
+getCacheDir :: Maybe FilePath -> IO (Maybe FilePath)
+getCacheDir damlPathM = do
+    dh <- lookupEnv damlCacheEnvVar
+    dir <- case dh <|> damlPathM  of
         Nothing -> fallback
         Just "" -> fallback
         Just var -> pure var
     createDirectoryIfMissing True dir
-    pure dir
+    perms <- getPermissions dir
```

Comment:
I’m slightly worried about the fact that we check this once here for the directory but don’t handle failures on the individual file writes and reads. That probably works in reasonable setups but I wouldn’t be surprised if we end up in situations where the directory is writable but individual files are not writable or not readable. Did you consider to instead guard the actual file operations and handle failures there? Afaict, we only have 4 kinds of file access:

1. `setOptIn`: If we cannot write `.opted_in` or remove `.opted_out` just do nothing.
2. `logOptOut`: If we cannot read the gcp metadata do nothing, if we cannot write the opted out file (we currently seem to do that asynchronously, that should probably change), do nothing.
3. `logMetadata`: If we cannot read the gcp metadata do nothing.
4. `sendData`: If we cannot access the sent data file, do nothing.

We can still keep the check here, if the directory is not writable there is no point in turning telemetry on but this should be a bit more robust in weird setups.

Reply:
Yes, that was my first approach but it became quite a mess. The idea is that it's our code generating these files. If the directory is not writable in the first place, we won't generate the files either. If we can create them, then they will be writable. If someone changes the permissions of these files afterwards, I think failing is acceptable.

Reply:
Alright, that seems convincing. Do we have to create this directory? I know that previously we didn’t have to create it because the assistant created it but now that it’s a separate directory, I’m not entirely clear if we can rely on it existing. This check probably then also needs to take account the directory not existing/run after we’ve created it.

Reply:
We create the directory if it doesn't already exist in `getCacheDir` in the gcp logger.